lib/update.go

```diff
@@ -0,0 +1,399 @@
+	go func() {
+		if err := svc.ServeHTTP(updateCfg.Address); err != nil {
+			log.Errorf("starting cron http server: %s", err)
+		}
+	}()
+
+	return svc.Start(ctx)
+}
+
+// ServiceStart ensures the scheduler is running
+func (m *UpdateMethods) ServiceStart(in, out *bool) error {
+	local, ok := m.inst.cron.(*cron.Cron)
+	if !ok {
+		return fmt.Errorf("service already running")
+	}
+
+	ctx := context.Background()
+	if err := local.Start(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ServiceStop halts the scheduler
+func (m *UpdateMethods) ServiceStop(in, out *bool) error {
+	// TODO (b5):
+	return fmt.Errorf("not finished")
+}
+
+// ServiceRestart uses shell commands to restart the scheduler service
+func (m *UpdateMethods) ServiceRestart(in, out *bool) error {
+	// TODO (b5):
+	return fmt.Errorf("not finished")
+}
+
+// Run advances a dataset to the latest known version from either a peer or by
+// re-running a transform in the peer's namespace
+func (m *UpdateMethods) Run(p *Job, res *repo.DatasetRef) (err error) {
+	if m.inst.rpc != nil {
+		return m.inst.rpc.Call("UpdateMethods.Run", p, res)
+	}
+
+	switch p.Type {
+	case cron.JTDataset:
+		params := &SaveParams{
+			Ref: p.Name,
+		}
+		if o, ok := p.Options.(*cron.DatasetOptions); ok {
+			params = &SaveParams{
+				Ref:                 p.Name,
+				Title:               o.Title,
+				Message:             o.Message,
+				Recall:              o.Recall,
+				BodyPath:            o.BodyPath,
+				FilePaths:           o.FilePaths,
+				Publish:             o.Publish,
+				Force:               o.Force,
+				ConvertFormatToPrev: o.ConvertFormatToPrev,
+				ShouldRender:        o.ShouldRender,
+				Secrets:             o.Secrets,
+
+				// TODO (b5) not fully supported yet:
+				// Strict: o.Strict,
+				// Config: o.Config
+			}
+		}
+		*res = repo.DatasetRef{}
+		err = m.runDatasetUpdate(params, res)
+
+	case cron.JTShellScript:
+		return base.JobToCmd(m.inst.streams, p).Run()
+
+	default:
+		return fmt.Errorf("unrecognized update type: %s", p.Type)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// TODO (b5): expand event logging interface to support storing additional details
+	return m.inst.Repo().LogEvent(repo.ETCronJobRan, *res)
+}
+
+func (m *UpdateMethods) runDatasetUpdate(p *SaveParams, res *repo.DatasetRef) error {
+	ref, err := repo.ParseDatasetRef(p.Ref)
+	if err != nil {
+		return err
+	}
+
+	if err = repo.CanonicalizeDatasetRef(m.inst.node.Repo, &ref); err == repo.ErrNotFound {
+		return fmt.Errorf("unknown dataset '%s'. please add before updating", ref.AliasString())
+	} else if err != nil {
+		return err
+	}
+
+	if !base.InLocalNamespace(m.inst.Repo(), &ref) {
+		*res, err = actions.UpdateRemoteDataset(m.inst.Node(), &ref, true)
+		return err
+	}
+
+	// default to recalling transfrom scripts for local updates
+	// TODO (b5): not sure if this should be here or in client libraries
+	if p.Recall == "" {
+		p.Recall = "tf"
+	}
+
+	saveParams := &SaveParams{
```

Comment:
Why copy?

Reply:
this was code left in place from the Save/UpdateParams refactor. nice catch